### PR TITLE
CustomGradientPicker: Update Type and Angle controls

### PR DESCRIPTION
**Pull request by ItsJonQ**
Screen Shot 2020-07-08 at 3 59 49 PM

This update improves the UI for the Linear/Gradient selector and angle controls for the `<CustomGradientPicker />` component from `@wordpress/components` (which is featured in the **Cover** block).

This helps partially resolve some of the suggestions proposed https://github.com/WordPress/gutenberg/issues/21269 by folks like @pablohoneyhoney, @jasmussen, and @mapk.

#### Linear/Radial select

The internal `<SelectControl />` has been updated with the new Gutenberg styles. The component functionality (that is the component API) remains the same. Since the UI of the new `<SelectControl />` matches that of `<InputControl />`, some of the internals were refactored so that they could be shared.

#### Angle control

These have also received a UI update to match Gutenberg styles. The functionality remains mostly the same, with some minor improvements to the drag interaction. Also, the input that appears next to the circle is updated with our new `NumberControl`.

86052782-84a23280-ba25-11ea-9161-7b3ed1a07320

## How has this been tested?

* Tested locally in Storybook and Gutenberg
* Run `npm run dev`
* Add a Cover block
* Select a background color
* Change it to Gradient
* Play with the Gradient controls

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [n/a] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by ItsJonQ** [comment]
Note: On the original issue, there were some interaction suggestions, such as...

> Worth looking at proportions (perhaps taller dragging area so it becomes hierarchically more prominent with a bigger area of interaction) and more distance between the slider and the controls below.

from @pablohoneyhoney 

> Can we allow this? It seems like a fair use case. Or even if I wanted to flip the gradient entirely, move the orange color all the way to the right past the red color?

from @jasmussen 

I took a look. It looks like it would involve some extensive refactors in the gradient slider portion of the component.

For now, I think it would be easier (at least to start) to focus on the controls that appear on the bottom

**Comment by jasmussen** [comment]
>  It looks like it would involve some extensive refactors in the gradient slider portion of the component.
> For now, I think it would be easier (at least to start) to focus on the controls that appear on the bottom

All good. I just saw some really sharp diagonal gradients that Kjell made, where there's no gradation because both colors are at the same position — so I ticketed it in #23816. Thank you for looking! 

Looking at this PR, here's a before:

Screenshot 2020-07-09 at 10 13 13

And here's after:

after

This is night and day, it is _profoundly_ better. It is such an improvement. 👏 Well done. 

There are a number of little followups I'd like to do separately, like putting all color swatches into a collapsible line line, revisit the clear buttons in a systematic way, and overall refine the sidebar. That is ongoing work, not forgotten:

Color controls, opens inline

... but something I will help with and which should not block this.

In other words, this is good to ship, Q, just needs a code sanity check! 👍 👍 — Thanks again.

**Comment by ItsJonQ** [comment]
> There are a number of little followups I'd like to do separately, like putting all color swatches into a collapsible line line, revisit the clear buttons in a systematic way, and overall refine the sidebar

@jasmussen Ah yes! I submitted a proposal to handle grouping/collapsing of like-controls:
https://github.com/WordPress/gutenberg/issues/23486

Thank you for your review + testing!

**Comment by munirkamal** [comment]
Hi @ItsJonQ 

As you are working on this component, what do you think about my proposal #24049 to add a radial direction option as well? That would enable some more design possibilities.

**Comment by ItsJonQ** [comment]
@munirkamal Hiya! Thanks for reaching out! I'll reply in your issue 🙏

**Comment by richtabor** [comment]
Love this 👏

**Review by ntsekouras — COMMENTED** [review]
Cool stuff here @ItsJonQ ! 💯  - I've left some small comments but most importantly I'd like you to share your opinion about `css-in-js`.

I had to read a bit about this to review your code :), so I'm just trying to better understand the trade offs and I know you've taken a deep dive in styles.

My first concerns are:
- if we're adding a new way of styling ( `emotion` `css-in-js` ) and it's trade offs have been considered, is it the way to go for other components ( new ones or refactored) as well?
- Should we use it only in complex scenarios? That means for example state based styles etc..

Thanks!

File: packages/components/src/input-control/input-base.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import Label from './label';
+import {
+	Container,
+	Root,
+	Prefix,
+	Suffix,
+	LabelWrapper,
+} from './styles/input-control-styles';
+
+function useUniqueId( idProp ) {
```
[the remark this answers is not in the document]
Comment (ItsJonQ) [reply packages/components/src/input-control/input-base.js:20]: @ntsekouras That's a good idea 👍 . Maybe something to do as a follow up!

File: packages/components/src/angle-picker-control/styles/angle-picker-control-styles.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+/**
```
[the remark this answers is not in the document]
Comment (ItsJonQ) [reply packages/components/src/angle-picker-control/styles/angle-picker-control-styles.js:1]: Good point! It may no longer be necessary. I originally introduced this structure to make it distinctly different from the `style.scss` during the early proposal/experimental phase.

I'm happy to keep the `style.js` files at the root level. It's the same convention I've followed in another React component library project:

https://github.com/ItsJonQ/g2/tree/master/packages/components/src/Button

File: packages/components/src/custom-gradient-picker/index.js [hunk 3]
```diff
@@ -71,27 +67,22 @@ const GradientTypePicker = ( { gradientAST, hasGradient, onChange } ) => {
...
-					{
-						icon: <RadialGradientIcon />,
-						title: __( 'Radial Gradient' ),
-						isActive: hasGradient && type === 'radial-gradient',
-						onClick: onSetRadialGradient,
-					},
-				] }
-			/>
-		</BaseControl>
+		<SelectControl
```
[the remark this answers is not in the document]
Comment (ItsJonQ) [reply packages/components/src/custom-gradient-picker/index.js:85]: Ah! Thank you for catching that. I've added the label back in via the `label` prop:

```jsx
<SelectControl
    className="components-custom-gradient-picker__type-picker"
    label={ __( 'Type' ) }
    onChange={ handleOnChange }
    options={ GRADIENT_OPTIONS }
    value={ hasGradient && type }
/>
```

It handles the `VisuallyHidden` based rendering + aria linking underneath 👍

File: packages/components/src/custom-gradient-picker/index.js [hunk 4]
```diff
@@ -71,27 +67,22 @@ const GradientTypePicker = ( { gradientAST, hasGradient, onChange } ) => {
 		);
 	};
 
+	const handleOnChange = ( next ) => {
```
[the remark this answers is not in the document]
Comment (ItsJonQ) [reply packages/components/src/custom-gradient-picker/index.js:75]: That's a good suggestion! I'm not sure how many cases we'll have in the near term. I think the simpler logic+setup is okay for now 😊

File: packages/components/src/angle-picker-control/styles/angle-picker-control-styles.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ntsekouras) [reply packages/components/src/angle-picker-control/styles/angle-picker-control-styles.js:1]: 💯

**Comment by aristath** [comment]
I just saw #23816 and was trying some things locally before coming here.
Changing the value of `MINIMUM_DISTANCE_BETWEEN_POINTS` in [`constants.js`](https://github.com/WordPress/gutenberg/pull/23802/files#diff-6b9d6354b2c5a5a40652de22e098cfca) from `9` to `1` allows points to overlap :+1:

**Comment by ItsJonQ** [comment]
@aristath Oo! Thanks for that suggestion. I just gave it a shot.

It definitely allows for the points to be brought closer. However, the UI controls feel awfully cramped when this happens 🙈 

Screen Shot 2020-08-11 at 1 10 57 PM

This feels especially cramped when there's the (+) inserted that appears on hover:



@jasmussen What do you think?

**Comment by jasmussen** [comment]
> Changing the value of MINIMUM_DISTANCE_BETWEEN_POINTS in constants.js from 9 to 1 allows points to overlap 👍

> @jasmussen What do you think?

Here's what I see.

gradient

I LOVE the ability to make those sharp gradients.

However in the above I actually set `MINIMUM_DISTANCE_BETWEEN_POINTS = 0;`, without that you can't make perfectly sharp lines like in the GIF. That, as you note, also makes the + behave a little funkily.

Which all is to say, I think it's a followup PR that should not block _this PR_ from happening, and in that followup we could probably also make it so the plus button does not appear when you mouse-over one of the existing handles. Maybe?

**Comment by mtias** [comment]
Great to see the progress here!

**Comment by jasmussen** [comment]
Screenshot 2020-08-27 at 12 13 02

Fixed size of bar to match mockups.

**Comment by jasmussen** [comment]
Harmonized spacing, and improved focus.

Screenshot 2020-08-27 at 12 28 19

**Comment by jasmussen** [comment]
Fixed so the swatches align on the right side with the gradient creator below.

Screenshot 2020-08-27 at 12 33 12

**Review by jasmussen — COMMENTED** [review]
This is a **vast** improvement over what's shipping. And after the recent polish, this is good enough to ship from a design point of view. 

I would love to get this in soon, so all it needs now is a final code review. @ntsekouras if you have the bandwidth, would appreciate you could take a look so we can get this in!

**Review by ntsekouras — APPROVED** [review]
Great work here @ItsJonQ and the polishing from @jasmussen 💯

**Comment by mtias** [comment]
Should we move the "type" label to above now or later?

image

**Review by mtias — APPROVED** [review]

**Comment by ItsJonQ** [comment]
Thanks for the feedback + enhancements all! I'll merge this up 🎉
